### ISSUE TITLE
Word Sense Disambiguation Module

### ISSUE THREAD

**Issue by Blubberli** [Blubberli/ambiguous-alpaca#75]
Idea: We are planning to evaluate / test our representations on whether they are useful for word sense disambiguation. 

Method: measure the cosine similarity between the representation of the phrase and each sense representation and pick the sense with the highest similarity.

to Do:

- *adj-noun-representations* : This is a representation for the adjective-noun phrase

1. `static`: method to load a composition model and generate and save predictions for the WSD adj-n phrases (we could import this method from joint_training.py and run it on a Dataloader which we construct for the WSD dataset)
2. `contextualized`: use the "get_single_word_embedding" method to retrieve an embedding for modifier and head and then compute the centroid

- *germanet-sense-representations* : This is a representation for GermaNet Sense ( = Synset)
I adapted the code to work for GermaNet and to generate .fifu embeddings:
https://github.com/mfaruqui/retrofitting
The paper for retrofitting:
https://arxiv.org/abs/1411.4166

1. `static`: use retrofitting to generate retrofitted sense embeddings for each GermaNet lexical unit, initialized with the same vector space as used to train the composition model. generate synset embeddings by computing the centroid and normalizing it
2. `contextualized`: use retrofitting to generate retrofitted sense embeddings for each GermaNet lexical unit, initialized with the German Bert vector space. An embedding for each word has to be computed by taking the centroid of all corresponding word piece embeddings. These embeddings will not be contextualized. generate synset embeddings by computing the centroid and normalizing it

- *wiktionary-paraphrase-representation*: This is an (optional) representation of the sense definition if available. 

Method: Let's have a look at this paper:
https://arxiv.org/abs/1907.13237
I re-read the paper and I guess I confused something. They do not use any special averaging method but they use all definition words.

1. `static`: centroid of definition words 
2. `contextualized`: centroid of definition words

- [optional]*extended context representation*: some representation of context sentence, e.g. centroid. (we currently use the adj-n phrase as context). could be added to the adj-n phrase to form one representation or we can compute the similarities between context and senses as well and then combine the results (we can also weigh the result based on the context and based on the adj-n phrase to give more weight to the adj-n phrase representation is we are mostly interested in whether we can use that for disambiguation

- *disambiguation*: compute the similarity between adj-noun-representation and each possible germanet-sense-representation for the ambiguous adjective (and if available the wiktionary-paraphrase-representation). average the similarities and predict the sense with the highest. compute accuracy
